YTEP-0036.rst: converting from nose to pytest

jcoughlin11 opened the pull request:
This PR covers the addition of YTEP-0036.rst. This new YTEP discusses changing yt's testing framework from nose to pytest, changing the testing architecture to remove yield tests, and using stored hashes of answer arrays instead of storing the entire answer array itself. In addition, it also details how to run the tests under pytest and how to write new answer tests in the new framework.

Commit:
Added YTEP-0036.rst, which details converting from nose to pytest and moving from stored arrays to stored array hashes.

ngoldbaum commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,261 @@
+
+.. code-block:: bash
+
+  $ pytest
```
Does this still only run the unit tests or does it run the answer tests too?

ngoldbaum replied:
Ah I see below it doesn't. Maybe note that here explicitly?

jcoughlin11 replied:
Done.

ngoldbaum commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,261 @@
+        for d in dobjs:
+          hashes['field_values'][f][d] = utils.generate_hash(self.field_values_test(ds, f, d))
+      hashes = {'test_method1': hashes}
+      utils.handle_hashes(self.save_dir, self.answer_file, hashes, self.answer_store)
```
It seems like a lot of this code is going to get repeated across all the answer tests in all the frontends. Might it be possible for the `AnswerTest` class to handle allocating the `hashes` dict, hashing the data, and storing the hashes for each test?

ngoldbaum reviewed (commented):
Overall this is really nice, thank you for working on this. This will make it much easier for people to run answer tests locally. See my comment below about reducing boilerplate - I think we want to make it as easy as possible to write new answer tests and I think it's worthwhile to think about insulating test authors from having to know about how hashes should be computed (and possibly screwing that up).

One of the nice aspects of the old answer testing system - especially for the plot answer tests - is that we had rich comparisons for the old answers and the new answers. For plot answer tests we even use matplotlib's image comparison machinery to generate a diff image which people can use to help figure out what exactly is changing, this is very useful in debugging. For the GitHub CI, it also generates a little HTML document that you can download as a test artifact that includes the before, after, and diff image.

Is that maintained here? I would guess that since we'd only be storing the image hashes we wouldn't be able to do that anymore. It might be nice to think about ways to continue storing the images - we're already putting them in the repo as-is (although with this it might make sense to offload the data storage to a secondary repo that only stores the images so we don't continue increasing the size of the yt repository).

In general it would be nice to have more discussion here in this document about the plot answer tests, since those are a little special and also got a lot of love during the google summer of code 2018.

The before/after comparison issue (e.g. a test fails and a developer wants to quickly understand what exactly changed) is less of a concern for the field values tests because we never really did anything to make it easy for developers to see exactly what changed between the new and old answers. That said, it might be nice to think about ways to ergonomically tell developers how to investigate failing answer tests, e.g. by somehow printing out a test script as well as git hashes to run the test script on to see the before/after values and before/after hashes explicitly. I could imagine some hacky ways to do this (e.g. by inspecting the source code for the test functions) but don't really have any concrete ideas - just a thought!

matthewturk commented:
@jcoughlin11 I agree with what @ngoldbaum said about the image tests in particular, but I wonder if we could also extend this somewhat to all of the tests.  As it stands, we are just storing the hashes, but most of the hashing is computed in a central location. Would it be possible to supply *arrays* to that central location, and have an option that stores the array values (even just in raw binary!) as well as checks the hashes?  That might help preserve both options.

It would also open up the opportunity for someone who is making changes and finding test errors to run the tests with that option (`--store-values` or something) and then load them in to manually compare in detail against the new tests that are failing.

matthewturk commented:
(Of course this could then be extended to save actual image files, etc etc.)

munkm commented:
I've been thinking a bit more about this (specifically related to the image diffs and how nice the gsoc work is that supports it). I agree that we should do our best to preserve that functionality, and ideally make sure the upload feature of the diff and the new data is preserved in CI as well. I'm not exactly sure how much more difficult it would be to supply arrays to the central location, but what if we had an option that if a test fails with the hash it then generates the diff and the image files? We could minimize storage needs by only generating these diffs if the test fails, but still make it so we can quickly identify issues with new features. 

One downside to this would be that if this is written where a test requires a rerun to generate the diffs, we could end up wasting a lot of resources running and re-running tests. But maybe that won't be an issue at all.

ngoldbaum commented:
One thing that hasn't come up yet that might be relevant for this discussion is that github actions are a thing now and they have [generous free compute resources](https://help.github.com/en/articles/about-github-actions#usage-limits) associated with them.

Commit:
Updated ytep-0036 to reflect the new changes made with regards to parametrizing tests, no longer using OrderedDicts, and reworking the hashing fixture.

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+
+Currently, testing in yt makes use of the `nose <https://nose.readthedocs.io/en/latest/>`_ framework. While perfectly functional, Nose has several drawbacks. First, it has been in a self-described maintenance mode for the last several years, which puts the project in danger of ceasing. Second, much of Nose's functionality comes from third-party plugins, which often have lacking or poor documentat ...
+
+The first proposal of this YTEP is to switch yt's testing framework from Nose to `pytest <http://pytest.org/en/latest/>`_. Pytest offers many of the same benefits of nose: automatic test discovery, the ability to selectively run tests, a large number of external plugins, the ability to be fine-tuned via configuration files, and compatibility with unittest. In addition to these benefits, pytest is ...
```
```suggestion
The first proposal of this YTEP is to switch yt's testing framework from Nose to `pytest <http://pytest.org/en/latest/>`_. Pytest offers many of the same benefits of nose: automatic test discovery, the ability to selectively run tests, a large number of external plugins, the ability to be fine-tuned via configuration files, and compatibility with `Python's standard library testing framework unittest <https://docs.python.org/3/library/unittest.html#module-unittest>`_. In addition to these benefits, pytest is also: actively maintained and developed, compatible with nose testing frameworks, and equipped with a fully-featured fixture system. In fact, this fixture system is arguably the best reason to use Pytest.
```

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+The second proposal of this YTEP is a general reorganization and streamlining for yt's answer testing framework.
+
+Currently, many answer tests in yt generate large arrays of data. In order to facilitate comparison, these arrays need to be saved. Unfortunately, their size necessitates that these answers be stored separately from the main yt code base and makes answer comparisons slow.
+
```
```suggestion
Synchronising pull-requests merging with two repos instead of one also slows down the development itself and creates technical debt.

```

neutrinoceros replied:
@jcoughlin11 this comment is still not marked "outdated". Maybe it simply doesn't fit the doc anymore, in which case you can simply resolve this.

neutrinoceros replied:
Oh wait it still applies. You can reject it if you want, no hard feelings.

jcoughlin11 replied:
I added your suggestion as the third bullet point here. I can make it its own paragraph if you think that would fit better.

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+
+Currently, the answer tests are mostly implemented as yield tests using nose. Pytest no longer supports yield tests because they resulted in multiple test functions being generated (one for each parameter combination), and these generated test functions did not properly support fixtures, as described `here <https://docs.pytest.org/en/latest/deprecations.html#yield-tests>`_.
+
+The first step in reorganizing the tests is, therefore, to remove all yield tests. Each function that previously yielded a test now properly calls a test function with the desired parameters, where assertions are made. This makes the tests easier to read, as most of the boilerplate associated with yielding is gone (e.g., changing the ``__name__`` attribute of the yielded test class).
```
>  Each function that previously yielded a test now properly calls a test function with the desired parameters

I'm confused, it sounds more like an intermediate step than a desirable final state: shouldn't test function be self-contained ? I believe that test generation is roughly equivalent to `@pytest.mark.parametrize`, is there something I'm missing ?

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+
+The first step in reorganizing the tests is, therefore, to remove all yield tests. Each function that previously yielded a test now properly calls a test function with the desired parameters, where assertions are made. This makes the tests easier to read, as most of the boilerplate associated with yielding is gone (e.g., changing the ``__name__`` attribute of the yielded test class).
+
+The second step is to organize the answer tests into classes (e.g., all of the enzo answer tests go in a class called ``TestEnzo``). This not only provides a natural organizational structure within the code itself, but it also provides a simple way of ensuring that all related answer tests get saved to the same answer file. The class structure also facilitates easier running of specific tests, es ...
```
> all related answer tests get saved to the same answer file

why is this important ?

> The class structure also facilitates easier running of specific tests

I don't really get this last argument. pytest offers several ways to select a subset of test at runtime already based on names (`pytest -k`) or markers (`pytest -m`). Why do you think it's not enough ?

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+
+The second step is to organize the answer tests into classes (e.g., all of the enzo answer tests go in a class called ``TestEnzo``). This not only provides a natural organizational structure within the code itself, but it also provides a simple way of ensuring that all related answer tests get saved to the same answer file. The class structure also facilitates easier running of specific tests, es ...
+
+As an extension of the above, each answer test (e.g., ``FieldValuesTest``) that was previously a class of its own is now a method (e.g., ``field_values_test``) of the base ``AnswerTest`` class, which resides in ``yt/utilities/answer_testing/framework.py`` and takes the place of the ``AnswerTestingTest`` class. The use of this base class provides a central location for any new answer tests that mi ...
```
I must be reading this wrong. Are you saying that those tests live in the `framework.py` ?

Upon reading a third time, I _think_ the source of my confusion is the phrasing of the first sentence. Specifically

> each answer test (e.g., ``FieldValuesTest``) 

I'm sure you mean something like "abstract answer test" here (I don't know of a better word here, sorry). In any case there's a distinction between generic/abstract tests and actual tests classes/functions, right ?
I think it should be clarified here and in the following.

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+
+As an extension of the above, each answer test (e.g., ``FieldValuesTest``) that was previously a class of its own is now a method (e.g., ``field_values_test``) of the base ``AnswerTest`` class, which resides in ``yt/utilities/answer_testing/framework.py`` and takes the place of the ``AnswerTestingTest`` class. The use of this base class provides a central location for any new answer tests that mi ...
+
+This structure removes the need for each individual answer test to have it's own ``__init__`` and ``compare`` methods. The test code is now exclusively what previously resided in the ``run`` method.
```
This isn't easy to follow without a minimal example of the before/after transformation, especially when the reader isn't familiar with the current (and soon, past) state of the framework.

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+
+This structure removes the need for each individual answer test to have it's own ``__init__`` and ``compare`` methods. The test code is now exclusively what previously resided in the ``run`` method.
+
+Several other changes have been implemented, as well. Previously, when running answer tests that used large data files, the ``--answer-big-data`` option had to be passed to nose. This remains the case, as described below. However, that option was previously passed to the decorators ``requires_ds`` and ``requires_sim``. This is no longer the case. Those decorators now only check to make sure the r ...
```
Could you please add a reference to https://github.com/yt-project/yt/pull/2885 here ?
(wow I now feel like a typical science-article reviewer "please cite me more" :) )

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+
+The last change made to the code base in switching from nose to pytest is the addition of ``conftest.py`` files.  These are configuration files that are used by pytest in order to define custom fixtures for processes such as setup, teardown, parametrizing tests, and using temporary directories and files.
+
+The primary ``conftest.py`` file resides in the root of the yt repo and it defines the command-line parser as well as the fixtures used across each of the answer tests. Additionally, other ``conftest.py`` files have been defined in each of the frontend directories in order to define fixtures for loading specific data files, and using ``pytest_generate_tests`` to parametrize those answer tests tha ...
```
```suggestion
The primary ``conftest.py`` file resides in the root of the yt repo and defines the command-line parser as well as the fixtures used across each of the answer tests. Additionally, other ``conftest.py`` files have been defined in each of the frontend directories in order to define fixtures for loading specific data files, and using ``pytest_generate_tests`` to parametrize those answer tests that require it.
```

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+Hash Implementation
+^^^^^^^^^^^^^^^^^^^
+
+In order to generate appropriate hashes, each method of ``AnswerTest`` that returns an array now has that array hashed via the ``hashing`` fixture defined in the central ``conftest.py`` file. This fixture saves the arguments that the test function employed (e.g., what field it was parametrized with) and then makes use of the ``md5`` method of the ``hashlib`` library to get the hash as a string of ...
```
While looking for a reference for this, I stumbled upon this https://docs.python.org/3/library/hashlib.html

> md5() is normally available as well, though it may be missing or blocked if you are using a rare “FIPS compliant” build of Python

But my main issue is that there's nothing else there :/
So my question is: why choose this particular hashing method ?

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+
+In order to generate appropriate hashes, each method of ``AnswerTest`` that returns an array now has that array hashed via the ``hashing`` fixture defined in the central ``conftest.py`` file. This fixture saves the arguments that the test function employed (e.g., what field it was parametrized with) and then makes use of the ``md5`` method of the ``hashlib`` library to get the hash as a string of ...
+
+In order to facilitate easier diagnosis of failing tests, each combination of test parameters receives its own hash. For example, if the ``field_values_test`` is being run on each field in ``ds.field_list``, then each call to ``field_values_test`` will generate its own hash. These hashes are then saved in a ``dict`` whose keys are the test parameter values (in this case, the field name). Once com ...
```
The end essentially reads "a dict whose keys are values", which is very confusing within the context of the Python language. Here's a rewording proposal

```suggestion
In order to facilitate easier diagnosis of failing tests, each combination of test parameters receives its own hash. For example, if ``field_values_test`` is being run on each field in ``ds.field_list``, then each call to ``field_values_test`` will generate its own hash. These hashes are then saved in a mapping of parameter names and values. Once completed, these are written to yaml files with the following format:
```

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+    test_parameter1_value1: value
+    test_parameter2_value1: value
+
+This produces small, human-readable text files that can be easily packaged with the main code base. This makes running and managing the tests simpler.
```
I dare say 😄 
```suggestion
This produces small, human-readable text files that can be easily packaged with the main code base. This makes running and managing the tests a lot simpler.
```

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+
+This produces small, human-readable text files that can be easily packaged with the main code base. This makes running and managing the tests simpler.
+
+The saving, loading, and comparison of the generated hashes has been implemented in a new convenience function ``handle_hashes(save_dir, save_file, hashes, answer_store)`` in ``yt/utilities/answer_testing/utils.py``.
```
`utils` gives basically no information on the content, which makes the code  unnecessarily harder to navigate _and_ to maintain. It opens the way for future additions that don't fit the original scope of the file, because it's not well reflected in it's name. This files has "utils" _and_ "utilities" as 2 of the 4 components of its path, and another one is the package name. (did I say I didn't like it ?)

Anyway this files predates the YTEP, but I'd like to point out that its name is not descriptive enough to my taste and I'd like to suggest we seize the opportunity to change it (or move your new stuff to somewhere else).

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+
+The ``save_dir`` argument is the path to the directory where the answers are either to be saved or loaded from, the ``save_file`` argument specifies the name of the answer file to either save to or load from, the ``hashes`` are the newly generated data to be either saved or loaded, and the ``answer_store`` argument specifies whether or not the passed data should be saved or compared.
+
+If ``answer_store`` is ``True`` then the passed ``hashes`` should be written to the specified yaml file. If ``answer_store`` is ``False`` then the passed ``hashes`` should be compared to the hashes loaded from the specified answer file.
```
which arguments are you talking about ? I fell like a sentence or two went missing. Also here a minimal example would help tremendously.

edit: after reading the following I think you meant command line flags, but I still don't get from the examples how the storage files are supposed to be specified. This should be clarified here.

edit2: ok so the storage (yaml) files are in fact configured in another yaml file, which belong to the repo ? In the context of command line argument/flags parsing, I think the word "configured" is less error-prone than "specified".

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+The last change made to the code base in switching from nose to pytest is the addition of ``conftest.py`` files.  These are configuration files that are used by pytest in order to define custom fixtures for processes such as setup, teardown, parametrizing tests, and using temporary directories and files.
+
+The primary ``conftest.py`` file resides in the root of the yt repo and it defines the command-line parser as well as the fixtures used across each of the answer tests. Additionally, other ``conftest.py`` files have been defined in each of the frontend directories in order to define fixtures for loading specific data files, and using ``pytest_generate_tests`` to parametrize those answer tests tha ...
+
```
I would like to propose to add a paragraph here on means to "test the tests", tools that we can use along the way to help validate the reimplementation of certain tests.

I think this plugin is an example of such tools : it shuffles the execution order of tests each time pytest is called, making sure that we don't accidentally and unknowingly end up writing tests that only happen to pass because of the specific order they are run in, which can happen when tests have global side effects (which can be unpredictable, hard to trace and not desired)
https://pypi.org/project/pytest-randomly/3.4.0/

I'm also thinking we could use coverage reports and badges to get a sense of progression going (even though I know we won't be converging towards 100%, it would still be very satisfying to see the pytest-coverage ratio being bumped with each PR, and let's not undermine the endorphin kick !)

(maybe this belongs to a different section, you decide)

jcoughlin11 replied:
I added a section discussing this. Let me know your thoughts!

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+
+  $ pytest
+
+Should one desire to disable pytest's default capturing of ``stdout``, the ``-s`` option can be invoked. If one would like a verbose description of the tests being run as well as their status (and a detailed report in the event of a failure), one can enable the ``-v`` flag.
```
```suggestion
Should one desire to disable pytest's default capturing of ``stdout``, the ``-s`` option can be invoked. If one would like a verbose description of the tests being run as well as their status (and a detailed report in the event of a failure), one can enable the ``-v`` flag (or ``-vv`` and ``-vvv`` for even finer reports).
```

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+
+.. code-block:: bash
+
+  $ pytest -s -v /path/to/test/module.py
```
remove the unnecessary, and show a realistic example of a test file as discovered by pytest
```suggestion
  $ pytest /path/to/tests/test_module.py
```

neutrinoceros replied:
updated this comment, still relevant @jcoughlin11

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+
+  $ yt config set yt test_data_dir /path/to/yt-data
+
+This remains true here, as well.
```
Not _exactly_ true anymore. Since https://github.com/yt-project/yt/pull/2981, it's now possible to have _local_ config files for yt. A side effect is that one can have as many data storage files as yt clones on a single machine.
However, the command line `yt` tool still only affects the global config file _unless specified otherwise_.
There no need to get into the details here but maybe a footnote could help.

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+
+.. code-block:: bash
+
+  nosetests --with-answer-testing --local --local-dir $HOME/Documents/test --answer-store --answer-name=local-tipsy yt.frontends.tipsy
```
```suggestion
  $ nosetests --with-answer-testing --local --local-dir $HOME/Documents/test --answer-store --answer-name=local-tipsy yt.frontends.tipsy
```

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+
+If one is writing tests for a new frontend, one should create a ``tests`` directory inside of the new frontend directory and, within the ``tests`` directory, create a file called ``test_outputs.py``. Within this file, one should define a new class that inherits from ``AnswerTest`` and is marked by pytest as being an answer test. All of the frontend tests should now be methods of this class. For e ...
+
+.. code-block:: python
```
note that this Python example code doesn't follow yt's code style (mostly black)
Fortunately, you can use this tool to get there in a second https://pypi.org/project/blacken-docs/

neutrinoceros replied:
@jcoughlin11 the test method still has a clunky indentation. Please resolve this thread when you're done updating it :)

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+
+If desired, each test method can be individually parametrized by creating a ``conftest.py`` file next to the ``test_outputs.py`` file. The example above would utilize this file like so:
+
+.. code-blocK:: python
```
```suggestion
.. code-block:: python
```

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+
+This ``pytest_generate_tests`` function will generate a separate test function for each combination of values in the two parameter lists. To parametrize against multiple parameters without having every combination:
+
+.. code-blocK:: python
```
```suggestion
.. code-block:: python
```

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+  # This is in yt/utilities/answer_testing.framework.py
+  class AnswerTest:
+    # The other methods are here. New method goes below
+    def my_new_answer_test(self, test_param1, test_param2):
```
what are the test discovery rules ? I'm wondering if `my_new_answer_test` is actually a "valid" test name according to pytest's default rules.

neutrinoceros replied:
The part of the doc this comment applied too has been removed, but my question hasn't been answered (unless the answer is in the new version, but I'm not looking at it right now. In this case, please resolve this thread @jcoughlin11).

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+      # Do stuff
+      return result
+
+Any new frontend must have its tests registered and assigned an answer file. This should be done by updating ``yt_repo/tests/tests.yaml`` with:
```
Hurray, that's a small doc PR we can do !

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,248 @@
+  TestNewFrontend:
+    answer_file_name_000.yaml
+
+Optionally, any new fixtures that you might define and use should go in the top-level ``conftest.py``, unless they are frontend specific, in which case they should go in a ``conftest.py`` file that resides in the ``frontends/my_new_frontend/tests`` directory.
```
```suggestion
Optionally, any new fixtures that one might define and use should go in the top-level ``conftest.py``, unless they are frontend specific, in which case they should go in a ``conftest.py`` file that resides in the ``frontends/my_new_frontend/tests`` directory.
```

neutrinoceros reviewed (commented):
This is excellent work. There's no question that I'm on board with the whole project but I listed a handful of questions I would like addressed, as well as a couple minor suggestions for improvements. Keep it up !

Commit:
Updated pytest ytep.

jcoughlin11 commented:
@neutrinoceros Thanks for the comments! I think I addressed all of them (let me know if that's not the case)! I also significantly reworked the document to reflect the infrastructure changes I made since this ytep was first written. I also removed a lot of the material that felt like a re-hashing of the pytest documentation.

neutrinoceros commented:
I tried resolving as many threads as I could @jcoughlin11, but I don't have time to read the whole new version again right now so I'll need to come back to this later. Thank you for simplifying the document, I look forward to proof reading again !

neutrinoceros commented on source/YTEPs/YTEP-0036.rst:
```diff
@@ -0,0 +1,223 @@
+        @pytest.mark.parametrize("param1", param1List, indirect=True)
+        @pytest.mark.parametrize("param2", param2List, indirect=True)
+        def test_method1(self, ds, param1, param2):
+          test_result = some_answer_test(ds, param1, param2)
+          self.hashes.update({'some_answer_test' : test_result})
```
you can resolve my comment about styling with this
```suggestion
            test_result = some_answer_test(ds, param1, param2)
            self.hashes.update({'some_answer_test' : test_result})
```

neutrinoceros reviewed (commented):
It looks even nicer than the last time I read it, well played !
I would like to add a small contribution that I think fits this very well and hasn't been discussed yet:
we have a lot of docstrings embedding example snippet suited for doctest, but we currently have no CI for those. Coincidently, it is extremely easy to make pytest run these along normal tests
(see https://docs.pytest.org/en/stable/doctest.html).
I expect this is a third subproject that can fit this YTEP. It will likely not be straightforward because many tests may be a little rotten, but I'm sure it's worth to do it.
Can you add a short paragraph (2 to 3 sentences should be enough ?) for this ?
Otherwise, after my remaining points have been adressed I'll be happy to approve this !

Commit:
Addresses review comments.

Commit:
Merge branch 'jcoughlin11/master'

Commit:
Added a brief section about doctests.

matthewturk commented:
So I *think* this is good to go.  I'm approving!  Anyone want to object to merging?

munkm commented:
I'm confused how all the unrelated YTEPS are touched in this PR? There should only be one file changed.

matthewturk commented:
@jcoughlin11 if you're OK with me pushing to your branch I can fix what @munkm mentions, I think.

Commit:
Merge branch 'master' of https://github.com/yt-project/ytep

matthewturk commented:
@munkm I think @jcoughlin11 fixed it with this

neutrinoceros commented:
> So I think this is good to go. I'm approving! Anyone want to object to merging?

✋🏻 I would like my last suggestion answered, even if not implemented.

jcoughlin11 commented:
@neutrinoceros Are you referring to your comment about doctest integration? If so, I added a section to the ytep called Doctest Integration that briefly covers this. Was there something else I was missing?

neutrinoceros commented:
@jcoughlin11  sorry my bad, I didn't see it (I was on my phone, it's harder to spot on a small screen)

Cool ! I figure we don't need to write down that the command line option `--doctest-glob` can be "adopted" in a config file, since it's not specific to this particular option.
However, could you also mention (or link to) this : https://docs.pytest.org/en/stable/doctest.html?highlight=doctest#doctest-namespace-fixture ? It seems like a pretty big deal to me since it allow to skip `>>> import yt` in doctest docstrings, which I remember adding in some places _just_ to make doctest happy.

Commit:
Adds a bit more to the doctest section.
